app/controllers/admin/deploy_groups_controller.rb

```diff
@@ -0,0 +1,52 @@
+class Admin::DeployGroupsController < ApplicationController
+  before_action :authorize_admin!
+  before_action :authorize_super_admin!, only: [ :create, :new, :update, :destroy ]
+  before_action :deploy_group, only: [:edit, :update, :destroy]
+
+  def index
+    @deploy_groups = DeployGroup.all
+  end
+
+  def new
+    @deploy_group = DeployGroup.new
+  end
+
+  def create
+    @deploy_group = DeployGroup.create(deploy_group_params)
+    if @deploy_group.persisted?
+      redirect_to action: 'index'
+    else
+      flash[:error] = "Failed to create deploy group: #{@deploy_group.errors.full_messages}"
+      render 'new'
+    end
+  end
+
+  def update
+    if deploy_group.update_attributes(deploy_group_params)
+      flash[:notice] = "Successfully saved deploy group: #{deploy_group.name}"
+      redirect_to action: 'index'
+    else
+      flash[:error] = "Failed to update deploy group: #{deploy_group.errors.full_messages}"
+      render 'edit'
+    end
+  end
+
+  def destroy
+    deploy_group.soft_delete!
+    flash[:notice] = "Successfully deleted deploy group: #{deploy_group.name}"
+    redirect_to action: 'index'
+  end
+
+  private
+
+  def deploy_group_params
+    params.require(:deploy_group).permit(:name, :environment_id)
+  end
+
+  def deploy_group
+    @deploy_group ||= DeployGroup.find(params[:id])
+  rescue ActiveRecord::RecordNotFound
```

Comment:
this can also be counter productive, for example when I type something by hand (the only real usecase except for bookmarks) I get redirected to a page I did not want to go to and have to re-type the whole url ... 
basically it's breaking the established 404 behavior that every other page uses for minimal gain and added code/test cost ... so imo not worth it